### PR TITLE
[opencascade] Initial port

### LINKED ISSUE #4968: Add OpenCASCADE
It would be great to have a port of the Open CASCADE library ( https://dev.opencascade.org/ ).

It is interesting to note that there is both the official code release of Open CASCADE, available at https://dev.opencascade.org/ and a community mantained version at https://github.com/tpaviot/oce . 
Debian currently packages both, while Homebrew packages the official one. 

Build instructions: 
 * https://dev.opencascade.org/doc/overview/html/occt_dev_guides__building_cmake.html .

Debian package : 
 * https://packages.debian.org/source/sid/opencascade .
 * https://packages.debian.org/source/sid/oce .

Homebrew formula : 
  * https://formulae.brew.sh/formula/opencascade .

Repology page:
  * https://repology.org/metapackage/opencascade/versions

### PR DESCRIPTION
**Pull request by marekr**
This adds OpenCascade / OCCT to resolve https://github.com/microsoft/vcpkg/issues/4968

A library intended for use with 3D CAD related applications and licensed under LGPL.

### CODE REVIEW

**Comment by JackBoosY** [comment]
Thanks for this PR! But we already have a PR to add this port https://github.com/microsoft/vcpkg/issues/4968.

File: ports/opencascade/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,68 @@
+vcpkg_fail_port_install(ON_TARGET "uwp")
```
Comment (PhoebeHui) [inline ports/opencascade/portfile.cmake:1]: From the 'Supports', it looks not support osx, linux and arm, could you add them to vcpkg_fail_port_install as well?

Could you aslo add 'vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)'?

File: ports/opencascade/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    REF V7_4_0
+    SHA512 595ad7226b9365c1a7670b77001f71787a5d8aaa4a93a4a4d8eb938564670d79ae5a247ae7cc770b5da53c9a9f2e4166ba6e5ae104c1f2debad19ec2187f4a56
+    HEAD_REF master
+)
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "dynamic")
+    set(LIBRARY_TYPE "Shared")
+else()
+    set(LIBRARY_TYPE "Static")
+endif()
```
Comment (PhoebeHui) [inline ports/opencascade/portfile.cmake:15]: Since the it doesn't support the static build, I don't think this is needed.

**Review by PhoebeHui — CHANGES_REQUESTED** [review]
@marekr, thanks for the PR!

File: ports/opencascade/CONTROL [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+Build-Depends: freeimage
+Description: Enable optional usage of freeimage
+
+Feature: gl2ps
+Build-Depends: gl2ps
+Description: Enable optional usage of gl2ps
+
+Feature: tbb
+Build-Depends: tbb
+Description: Enable optional usage of tbb
```
[the remark this answers is not in the document]
Comment (marekr) [reply ports/opencascade/CONTROL:17]: It builds and links to them without issue. As for if they work is a slight unknown.

File: ports/opencascade/portfile.cmake (hunk 2 above)
Comment (marekr) [reply ports/opencascade/portfile.cmake]: Well, static libraries are supposed to be possible but weirdly it seems the cmake script is broken upstream and needs patching. I'll remove this stub.

File: ports/opencascade/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+file(GLOB BATS ${CURRENT_PACKAGES_DIR}/bin/*.bat)
+file(REMOVE_RECURSE ${BATS})
+
+# move cmake files into the share folder
+file(GLOB CMAKES ${CURRENT_PACKAGES_DIR}/lib/cmake/opencascade/*.cmake)
+foreach(CM ${CMAKES})
+    get_filename_component(N "${CM}" NAME)
+    file(RENAME ${CM} ${CURRENT_PACKAGES_DIR}/share/${PORT}/${N})
+endforeach()
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/lib/cmake)
```
Comment (PhoebeHui) [inline ports/opencascade/portfile.cmake:61]:
```suggestion
```

File: ports/opencascade/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        -DBUILD_MODULE_Draw=OFF
+        -DINSTALL_DIR_LAYOUT=Unix
+        -DBUILD_SAMPLES_MFC=OFF
+        -DBUILD_SAMPLES_QT=OFF
+        -DBUILD_DOC_Overview=OFF
+        ## Options from vcpkg_check_features()
+        ${FEATURE_OPTIONS}
+)
+
+vcpkg_install_cmake()
```
Comment (PhoebeHui) [inline ports/opencascade/portfile.cmake:38]:
```suggestion
vcpkg_install_cmake()

vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/opencascade TARGET_PATH share/opencascade)
```

**Review by PhoebeHui — COMMENTED** [review]
You can use vcpkg_fixup_cmake_targets to move the cmake config and targets files to share folder.

**Comment by marekr** [comment]
> You can use vcpkg_fixup_cmake_targets to move the cmake config and targets files to share folder.

Good to know. It appears to work fine.

**Comment by PhoebeHui** [comment]
I build opencascade:x86-windows with vs2019 16.7.2 manually, it hang when building with release configuration, I already set it 'DISABLE_PARALLEL_CONFIGURE', howvere it doesn't solve the issue. It's unclear for me now, I will double confirm if it's a VS bug.

**Comment by marekr** [comment]
I build using 16.7.2 without issue.

**Comment by marekr** [comment]
Hmm, I did a test compile of x86-windows. It does appear only the Release build ends up stuck. It's kind of funny, it ended up with four cl.exe instances each using exactly the same amount of CPU which is....interesting to watch

ys9DPwhAjh


Side note, I removed the gl2ps feature because I didn't realize it was removed in 7.3 >_< freeimage and tbb features are definitely still there.

**Comment by ericbrumer** [comment]
Hey @marekr I work on the Visual C++ team, and I can reproduce the issue that you're seeing. opencascade has exposed a bug in our improved inliner heuristics. The good news is that you can throw a switch to disable them until we're able to ship a fix. Just pass `/d2DeepThoughtInliner-` and the compile time problem should go away. I confirmed that on one of the samples where we were hanging.

We'll look into fixing the bug in a future release. Going forward, if you run into compiler problems feel free to create a ticket on [Developer Community](https://developercommunity.visualstudio.com/content/problem/post.html?space=62) . This particular bug would have been routed to my team.

Thanks!

**Comment by marekr** [comment]
Hmm, ok thanks @ericbrumer. Flag applied/committed and it compiles fine now @PhoebeHui

**Comment by PhoebeHui** [comment]
The failures with x86-windows would be fixed by PR https://github.com/microsoft/vcpkg/pull/13722. I will will rerun the CI testing after the PR merged.

File: ports/opencascade/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+)
+
+# VTK option in opencascade not currently supported because only 6.1.0 is supported but vcpkg has >= 9.0
+
+
+if (TRIPLET_SYSTEM_ARCH MATCHES "x86")
+	# Turn off the "improved inline analyzer" that stalls on x86 release builds using MSVC 16.7
+    string(APPEND VCPKG_C_FLAGS "/d2DeepThoughtInliner-")
+    string(APPEND VCPKG_CXX_FLAGS "/d2DeepThoughtInliner-")
+endif()
```
Comment (BillyONeal) [inline ports/opencascade/portfile.cmake:24]: Can this be patched in the inner build instead? This shouldn't be added for non-MSVC; x86-linux is a thing.

Comment (marekr) [reply ports/opencascade/portfile.cmake:24]: I can guard it with win32. It just seems far cleaner to simply pass in flags at the portfile level rather than creating an entire patch file that someone will have to maintain every version bump of opencascade

Comment (BillyONeal) [reply ports/opencascade/portfile.cmake:24]: Guarding it for win32 is wrong because clang on Windows exists. I know we prefer to do things in portfile.cmake rather than patches but in this case it really does need to target only MSVC and we don't know that it's MSVC in use in portfile.cmake

Comment (marekr) [reply ports/opencascade/portfile.cmake:24]: > I know we prefer to do things in portfile.cmake rather than patches but in this case it really does need to target only MSVC and we don't know that it's MSVC in use in portfile.cmake

CMake has a variable for that though if( MSVC )....it's literally the same check I need to put into the portfile and would affect Linux and clang builds as well?

Comment (marekr) [reply ports/opencascade/portfile.cmake]: @BillyONeal Patch added.

Comment (BillyONeal) [reply ports/opencascade/portfile.cmake]: >CMake has a variable for that though if( MSVC )....

That comes from `project(... CXX)` meaning it shouldn't be filled in when `portfile.cmake` is running.

Comment (marekr) [reply ports/opencascade/portfile.cmake]: I am starting to see what you mean

I have committed the desired patch to opencascade's source

**Comment by BillyONeal** [comment]
Thanks for your contribution!

**Comment by HaroldReyiz** [comment]
Hi, don't know if this is the right issue for this but, I just tried to vcpkg install opencascade and I got:
`-- Note: opencascade only supports dynamic library linkage. Building dynamic library.
CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:42 (message):
  Refusing to build unexpected dynamic library against the static CRT.  If
  this is desired, please configure your triplet to directly request this
  configuration.
Call Stack (most recent call first):
  ports/opencascade/portfile.cmake:2 (vcpkg_check_linkage)
  scripts/ports.cmake:136 (include)`

I'm using triplet x64-windows-static.

**Comment by BillyONeal** [comment]
@HaroldReyiz The port `opencascade` does not support static builds; the upstream library apparently only supports being built as a DLL.

**Comment by HaroldReyiz** [comment]
I vaguely remember there was a variable exposed for static building in CMake. So I thought it was possible to statically build OpenCascade via CMake.

**Comment by BillyONeal** [comment]
> I vaguely remember there was a variable exposed for static building in CMake. So I thought it was possible to statically build OpenCascade via CMake.

CMake has `BUILD_SHARED_LIBS` but a project can pass `STATIC` or `SHARED` to `add_library`, or might just not build successfully in such cases.

A library is likely to force DLL deployment if they have global variables that need to be consistent for a whole process.

**Comment by marekr** [comment]
Opencascade has a static library option but when I tried to use it I ended
up with all kind of errors that suspiciously seemed like nobody upstream
has cared to test it.
And I didn't quite need static builds myself to want to spend the effort
patching the problems and having dynamic builds for windows was more than
valuable enough to contribute.

**Comment by HaroldReyiz** [comment]
I understand. Thanks a lot for the contribution by the way!

**Comment by ahoarau** [comment]
@marekr Did you try to include tcltk 8.6 for the Draw option ?

**Comment by marekr** [comment]
> @marekr Did you try to include tcltk 8.6 for the Draw option ?

Only in so far as seeing vcpkg has a tcl 9.0 port that isn't compatible and no tk port. And after looking at the tcltk's lack of real documentation for the build process/application level instead of the language, I gave up. Supporting just C++ level builds was more than adequate to support use by the two major foss programs FreeCad (mechanical) and KiCad (electrical)

**Comment by Dejauxvue** [comment]
Hi,
your port builds fine for me. However something seems to go wrong during installation:
I run vcpkg.exe opencascade:x64-windows and have the following example:
<details>
<summary>CMakeLists.txt</summary>

```
cmake_minimum_required(VERSION 3.12 FATAL_ERROR)
project(test)
find_package(OpenCASCADE CONFIG REQUIRED)
add_executable(test "")

target_sources(
	test
	PRIVATE
		"main.cpp")
target_link_libraries(
	test
	PRIVATE
	TKBO
	TKBin
	TKCAF
	TKCDF)
```
</details>

and I run it with
<details>
<summary>CMakeSettings.json</summary>
```
{
  "configurations": [
    {
      "name": "x64-Debug (default)",
      "generator": "Ninja",
      "configurationType": "Release",
      "inheritEnvironments": [ "msvc_x64_x64" ],
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "variables": [],
      "cmakeToolchain": "<path>/vcpkg/scripts/buildsystems/vcpkg.cmake"
    },
    {
      "name": "x64-Release",
      "generator": "Ninja",
      "configurationType": "Release",
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "",
      "buildCommandArgs": "",
      "ctestCommandArgs": "",
      "cmakeToolchain": "<path>/vcpkg/scripts/buildsystems/vcpkg.cmake",
      "inheritEnvironments": [ "msvc_x64_x64" ],
      "variables": []
    }
  ]
}
```
</details>


on Visual Studio and get the following error:
<details>
<summary>Console Output</summary>
```
[CMake] CMake Error at <path>/vcpkg/installed/x64-windows/share/opencascade/OpenCASCADEFoundationClassesTargets.cmake:84 (message):
1> [CMake]   The imported target "TKernel" references the file
1> [CMake] 
1> [CMake]      "<path>/vcpkg/installed/x64-windows/debug/libd/TKernel.lib"
1> [CMake] 
1> [CMake]   but this file does not exist.  Possible reasons include:
1> [CMake] 
1> [CMake]   * The file was deleted, renamed, or moved to another location.
1> [CMake] 
1> [CMake]   * An install or uninstall procedure did not complete successfully.
1> [CMake] 
1> [CMake]   * The installation package was faulty and contained
1> [CMake] 
1> [CMake]      "<path>/vcpkg/installed/x64-windows/share/opencascade/OpenCASCADEFoundationClassesTargets.cmake"
1> [CMake] 
1> [CMake]   but not all the files it references.
1> [CMake] 
1> [CMake] Call Stack (most recent call first):
1> [CMake]   <path>/vcpkg/installed/x64-windows/share/opencascade/OpenCASCADEConfig.cmake:95 (include)
1> [CMake]   <path>/vcpkg/scripts/buildsystems/vcpkg.cmake:555 (_find_package)
1> [CMake]   CMakeLists.txt:5 (find_package)
1> [CMake] -- Configuring incomplete, errors occurred!
```
</details>

**Comment by ahoarau** [comment]
Actually I was planning to downgrade the provided TCL to the latest stable (8.6.11) and provide  a build for TK (8.6.11-1) but failed on the vcpkg_install_nmake() macro.

The difficulty comes from the fact that TK supports only 2 types of scenarios:
1. In tree build : 

```
cd C:\path-to-tcl-sources
nmake -nologo -f makefile.vc
cd C:\path-to-tk-sources
nmake -nologo -f makefile.vc install TCLDIR=C:\path-to-tcl-sources
```

In here the tk's rules.vc expects the source tree to have everything it needs (includes in generic/, somes compiles tcl libs in win/Debug_MSVC_2019 etc etc).


2. Install build

```
cd C:\path-to-tcl-sources
nmake -nologo -f makefile.vc install INSTALLDIR=C:\tcltk8.6
cd C:\path-to-tk-sources
nmake -nologo -f makefile.vc install INSTALLDIR=C:\tcltk8.6
```

In there tk's rules.vc expects the install tree to not be modified by vcpkg's standards (no weird debug/release sharing includes at different levels)

Basically you only need to execute 2 simples nmake commands to build tcl and tk, but somehow the equivalent CMake/vcpkg code is not that easy.
